src/outputs/rcp.c

```diff
@@ -0,0 +1,1464 @@
+
+static void
+rcp_status(struct rcp_session *s)
+{
+  enum output_device_state state;
+
+  switch (s->state)
+    {
+      case RCP_STATE_SETUP:
+      case RCP_STATE_SETUP_WAKEUP:
+	state = OUTPUT_STATE_STARTUP;
+	break;
+
+      case RCP_STATE_SETUP_GET_CONNECTED_SERVER ... RCP_STATE_SETUP_VOL_GET:
+      case RCP_STATE_QUEUING_CLEAR ... RCP_STATE_QUEUING_PLAY:
+      case RCP_STATE_VOL_GET:
+      case RCP_STATE_VOL_SET:
+      case RCP_STATE_STOPPING ... RCP_STATE_STANDBY:  // as per @frankusb, move advise that remote/we're not interested
+	state = OUTPUT_STATE_CONNECTED;
+	break;
+
+      case RCP_STATE_STREAMING:
+	state = OUTPUT_STATE_STREAMING;
+	break;
+
+      case RCP_STATE_DISCONNECTED:
+	state = OUTPUT_STATE_STOPPED;
+	break;
+
+      case RCP_STATE_FAILED:
+      default:
+	state = OUTPUT_STATE_FAILED;
+    }
+
+  DPRINTF(E_DBG, L_RCP, "mapping state from (internal) %d -> (output) %d\n", s->state, state);
+  outputs_cb(s->callback_id, s->device->id, state);
+  s->callback_id = -1;
+
+  if (state == OUTPUT_STATE_STOPPED || state == OUTPUT_STATE_FAILED)
+    rcp_session_cleanup(s);
+}
+
+
+/* ------------------ INTERFACE FUNCTIONS CALLED BY OUTPUTS.C --------------- */
+
+static int
+rcp_device_start(struct output_device *device, int callback_id)
+{
+  struct rcp_session *s;
+
+  DPRINTF(E_DBG, L_RCP, "device start %ld '%s'\n", device->id, device->name);
+
+  s = rcp_session_make(device, callback_id);
+  if (!s)
+    return -1;
+
+  return 1;
+}
+
+static int
+rcp_device_stop(struct output_device *device, int callback_id)
+{
+  struct rcp_session *s = device->session;
+
+  DPRINTF(E_DBG, L_RCP, "device stop %ld '%s'\n", device->id, device->name);
+
+  /* force these devices as deselected (auto state saves in db later) since 
+   * these need use to select (and cause the device probe to start connection to
+   * remote side
+   */
+  device->selected = 0;
+
+  s->callback_id = callback_id;
+  // tear this session down, incl free'ing it
+  rcp_session_shutdown_init(s);
+
+  return 1;
+}
+
+static int
+rcp_device_flush(struct output_device *device, int callback_id)
+{
+  DPRINTF(E_DBG, L_RCP, "device flush %ld '%s'\n", device->id, device->name);
+
+  // needed?
+  struct rcp_session *s = device->session;
+
+  s->callback_id = callback_id;
+  s->state = OUTPUT_STATE_STOPPED;
+
+  rcp_status(s);
+
+  return 1;
+}
+
+static int
+rcp_device_probe(struct output_device *device, int callback_id)
+{
+  struct rcp_session *s;
+
+  DPRINTF(E_DBG, L_RCP, "device probe %ld '%s'\n", device->id, device->name);
+
+  s = rcp_session_make(device, callback_id);
+  if (!s)
+    return -1;
+
+
+  return 1;
+}
+
+
+static int
+rcp_device_volume_set(struct output_device *device, int callback_id)
+{
+  struct rcp_session *s = device->session;
+  char cmd[4];
+  int ret;
+
+  DPRINTF(E_DBG, L_RCP, "device volume set %ld '%s' state=%d vol=%d\n", device->id, device->name, s->state, device->volume);
+
+  if (!s || (s && s->state != RCP_STATE_STREAMING))
```

Comment:
!s || (s && xxx) -> superfluous test for s

Reply:
protects for `s == NULL` or `s->state != RCP_STATE_STREAMING`.

I'm assumeing taht `device->session` could be `NULL` but if this is not the case then yes can drop `!s`

Reply:
The superfluous test is the s in (s && xxx). Logically, s is always non-null at this point, otherwise the part after || would not be evaluated.